<?xml version="1.0" encoding="UTF-8"?>
<launch>

  <!-- ROS PARAMETERS -->

  <!-- System level -->
  <arg name="namespace_tiago"     default="/"/>
  <arg name="namespace_project"   default="robotics_intro"/>
  <arg name="public_sim"          default="True"/>
  <arg name="sim_sufix"           value="_public_sim" if="$(arg public_sim)"/>
  <arg name="sim_sufix"           value=""  unless="$(arg public_sim)"/>
  <arg name="tf_prefix"           default=""/>
  <env name="PAL_HOST"            value="tiago" />
  <env name="LC_NUMERIC"          value="C" />

  <!-- Navigation -->
  <arg name="provide_map"     default="false"/>
  <arg name="localization"    default="amcl"/>
  <arg name="map_path"        default="$(find robotics_project)/pal/tiago_maps/config"/>
  <arg name="global_planner"  default="navfn"/>
  <arg name="local_planner"   default="eband"/>
  <arg name="planner"         default="base"/>
  <!-- Robot pose in Gazebo for initial demo -->
  <arg name="gzpose"         default="-x -0.3343 -y -0.0321 -z -0.0012 -R -0.005 -P 0.00322 -Y -0.8355"/>
  <!-- Robot pose in Gazebo for Tasks E and C -->
  <!-- <arg name="gzpose"         default="-x -1.146 -y -6.145 -z -0.001 -R -0.002 -P 0.001 -Y -1.593"/> -->
  <!-- Robot pose in Gazebo for Task A -->
  <!-- <arg name="gzpose"     default="-x 2.40 -y -11.0 -z -0.003 -R 0.0 -P 0.0 -Y 70.0"/> -->
 
  <!-- Simulation -->
  <arg name="robot"       default="steel"/>
  <arg name="rviz"        default="true"/>
  <arg name="gzclient"    default="true"/>
  <arg name="world"       default="appartment"/>
  <arg name="recording"   default="false"/>
  <arg name="multiple"    default="false"/>

  <!-- Definitions of pregrasp motion -->
  <rosparam command="load" file="$(find robotics_project)/config/pick_motions.yaml" /> 

  <!-- ROS NODES -->

  <group ns="$(arg namespace_tiago)">
    <!-- Gazebo simulation and TIAgo components -->
    <include file="$(find tiago_gazebo)/launch/tiago_gazebo.launch">
      <arg name="world"       value="$(arg world)"/>
      <arg name="robot"       value="$(arg robot)"/>
      <arg name="gzpose"      value="$(arg gzpose)"/>
      <arg name="gui"         value="$(arg gzclient)"/>
      <arg name="public_sim"  value="$(arg public_sim)"/>
      <arg name="recording"   value="$(arg recording)"/>
    </include>

    <!-- Manual controller -->
    <!-- <node pkg="key_teleop" type="key_teleop.py" name="manual_controller" output="screen" launch-prefix="xterm -e">
    </node> -->

    <!-- Project RGBD point cloud to laser-scan -->
<!--     <include file="$(find tiago_laser_sensors)/launch/rgbd_depth_laser.launch">
      <arg name="image"     value="xtion/depth_registered/image_raw"/>
      <arg name="scan"      value="rgbd_scan"/>
      <arg name="multiple"  value="$(arg multiple)"/>
      <arg name="tf_prefix" value="$(arg tf_prefix)"/>
    </include> -->

    <!-- AMCL Localization -->
    <!-- <include file="$(find tiago_2dnav)/launch/localization_$(arg localization).launch">
      <arg name="multiple"  value="$(arg multiple)"/>
      <arg name="tf_prefix" value="$(arg tf_prefix)"/>
    </include> -->

    <!-- Planning: move_base stack -->
    <!-- <include file="$(find tiago_2dnav)/launch/move_$(arg planner).launch">
      <arg name="global_planner" value="$(arg global_planner)"/>
      <arg name="local_planner"  value="$(arg local_planner)"/>
      <arg name="public_sim"     value="$(arg public_sim)"/>
      <arg name="tf_prefix"      value="$(arg tf_prefix)"/>
      <arg name="multiple"       value="$(arg multiple)"/>
    </include> -->
  </group>

  <group ns="$(arg namespace_project)">
    <!-- Upper-level state machine -->
    <node pkg="robotics_project" type="sm_students.py" name="logic_state_machine" output="screen">
      <param name="global_loc_srv" value=""/>
      <param name="clear_costmaps_srv"  value=""/>
      <param name="pick_srv"            value=""/>
      <param name="place_srv"           value=""/>
      <param name="nav_goal_topic"      value=""/>
      <param name="cmd_vel_topic"       value="/key_vel"/>
      <param name="move_base_frame"     value=""/>
      <param name="pick_pose_topic"     value=""/>
      <param name="place_pose_topic"    value=""/>
      <param name="aruco_pose_topic"    value=""/>
      <param name="cube_pose"           value="0.50306828716, 0.0245718046511, 0.915538062216, 0.0144467629456, 0.706141958739, 0.707257659069, -0.0306827123383"/>
      <param name="base_frame"          value="" />
      <param name="move_head_srv"       value="/$(arg namespace_project)/move_head_gui" />
    </node>

    <!-- Goal poses for navigation -->
<!--     <node pkg="robotics_project" type="poses_pub.py" name="poses_publisher" output="screen">
      <param name="pick_pose"         value="-1.1480, -6.1, -0.001, 0.0, 0.0, -0.709307863674, 0.70489882574"/>
      <param name="place_pose"        value="2.6009, -1.7615, 0.0, 0.0, 0.0, 0.0, 1"/>
      <param name="pick_pose_topic"   value="/place_holder1"/>
      <param name="place_pose_topic"  value="/place_holder2"/>
      <param name="map_frame"   value="/place_holder3"/>
    </node> -->

    <!-- Node exposing service to trigger the pick and place -->
    <node pkg="robotics_project" name="manipulation_client" type="manipulation_client.py" output="screen">  
      <param name="marker_pose_topic"   value="/place_holder4"/>
      <param name="pick_srv"            value="/place_holder5"/>
      <param name="place_srv"           value="/place_holder6"/>
      <param name="move_head_srv"       value="/$(arg namespace_project)/move_head_gui"/>
      <param name="pickup_marker_pose"  value="/place_holder7"/>
      <param name="place_marker_pose"   value="/place_holder8"/>
    </node>

     <!-- Arm manipulation node -->
    <node pkg="robotics_project" name="manipulation_server" type="manipulation_server.py" output="screen">
        <rosparam command="load" file="$(find robotics_project)/config/pick_and_place_params.yaml" />
        <param name="pickup_marker_pose"  value="/place_holder9" />
        <param name="place_marker_pose"   value="/place_holder10" />
    </node>

    <!-- Aruco marker detector -->
<!--     <node pkg="aruco_ros" type="single"  name="aruco_single" output="screen">
        <remap from="/camera_info"        to="/xtion/rgb/camera_info" />
        <remap from="/image"              to="/xtion/rgb/image_rect_color" />
        <param name="marker_pose_topic"   value="/place_holder11"/>
        <param name="base_frame"          value="/base_footprint"/>
        <param name="camera_frame"        value="xtion_rgb_optical_frame"/>
        <param name="marker_frame"        value="aruco_frame" />
    </node> -->
  </group>

  <!-- Map providers -->
<!--   <group if="$(arg provide_map)">
    <node name="map_server" pkg="map_server" type="map_server" args="$(arg map_path)/map.yaml"/>
    <node name="relay_map" pkg="topic_tools" type="relay" args="/map /vo_map"/>
  </group> -->

  <!-- RVIZ -->
  <node name="rviz" pkg="rviz" type="rviz" if="$(arg rviz)" args="-d $(find tiago_2dnav)/config/rviz/navigation$(arg sim_sufix).rviz"/> 

</launch>
